### PR TITLE
chore(*): Upgrade to hyper-0.14.x

### PR DESCRIPTION
**Pull request by jolhoeft**
At long last, we have an async version. This PR builds, and all but one example program works and have been tested (except the long broken https example). This is complete enough that the remaining work can be broken into more manageable pieces. There are some limitations:

* `cargo test` fails horribly. This PR is already large, and the integration tests require rewriting the test harness, so I plan on fixing the tests in the next PR. I could be persuaded otherwise though.

The following are all for a future PR.

* async functions don't yet have a default `Middleware` implementation. This should be mostly a case of figuring out the correct type signature.
* `middleware` and `router` macros do not support async. This is more complex to fix. We may want to combine this with a redesign of the `middleware` macro, which has confusing behavior. This is why the `macro_example` program is disabled.

These two items mean the only way to write an async middleware is to implement the `Middleware` trait. A number of the example programs demonstrate how to do this.

* `plugins` are disabled due to the plugin crate not not being able to be restricted to `Sync + Send`. The `plugin` crate may be abandoned, so fixing this may be a headache.

* `Response::on_send` is currently disabled. Feedback on its use cases is needed, in particular does it need to be async, which is more complex.

### CODE REVIEW

**Comment by GitCop** [comment]
Thanks for contributing! Unfortunately, I'm here to tell you there were style issues with your Pull Request:

 * Commit: 148f1131eef783bacd1b94382071995b2e5219a3
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 7f01f95114b56534316bae051c786bf1dfa4cdb0
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 4cd09ccd0e3e3c77413c1288bc48e127716fc0d4
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 264cd23e512606dc07d736702d588dc1c34a3883
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 362e4779517836c1b36011a9285d81b2999b0364
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 8a828acb8ddf17187059c1d8f9fd7fa6f0029e18
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 8293706cbc111da247e85143594b13885401a6b0
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: ea3d8f13c6bf6b6635178904ddaab52b14951c7f
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 8ff8981f43f73017083b4ad014909109085bf2d9
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: df8b94f35a2f70efae075eced8e6eef817a0e3e8
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: ad183a6ffbf97daeeac3b950b8571461a6c85956
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 6896d75e5193bfd7aac2655cfaf65e81ec16ffc8
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 63a7723cab97c4015239c1d83f5081547f6758da
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 4cb386ac04512e113d505b4bebd37ea0ec2780b7
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 84038c45d7a384f0a5141c78e54f2e15b6246ea6
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: ee96b9f4f9f0f3b030c61b81bf721c4688c39292
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 399da651c6c58261f8cc477fa132f9ae32261378
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 52ec773ed4f4ace52148602182c6ad26393eabd6
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 51c1f51aec5e3da0ab39028d4599d124e8b57c66
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 72409b21ef3c606d49a73b7807e0c5402645481d
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: e13cfe6bb3f073bb60362d44c51f86c1d8d3b56a
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 4bd354474499dd1df0aada3fc6af83892566c4b9
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: fe20927654a7958f56ce7ed5a1536d4542463900
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: e90ed6f5992ebad01ec46ae4032d0265d7cdc9e8
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 5a7524dd3d573da456da9eb37fef7a38b902c84e
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 75daf1bd34f9bde78957181f19fd3369a88dfd27
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**
  * Invalid type. Valid types are **feat, fix, docs, style, refactor, perf, test, chore, revert, doc**

 * Commit: 6495879c7378af46fcfc2f7a0634e9a0f2e71dff
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: ee8b8132d03980f4d1e3e237d72aa40ef934eeae
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: c00e64c7e1accccfdfe256d7bb302d4637b5d811
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 3c91d388e98dd6490851b94babfbd21a2df1e3b4
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/nickel-org/nickel.rs/blob/master/contributing.md

---
This message was auto-generated by https://gitcop.com

**Review by therealbstern — APPROVED** [review]
The only trouble spots I see, from most meaningful to least, are:
- there's a lot of TODO for this refactor still present
  - it's not clear that they're all addressed
- the loss of plugins is concerning
  - I'm especially concerned about `multipart`
  - I don't see `cookies`/`nickel-cookies` as a builtin either
- there's some redundant interfaces that should be dropped (`body_method`, for example)
- `listener.socket()` appears gone, and this is a mildly interesting method to keep
- `origin` has become `pub` which is probably fine but will be conceptually hard to make private again later if need be
- SSL support is moved into `hyper` or otherwise gone?
- various startup/shutdown interface stuff
  - `output_on_listen` tested and used in some places where it's not actually making anything clearer
  - `output_on_listen` not being used in some places where the listen socket is printed
  - `end` is now a no-op

Overall, it looks good to me, though this would benefit from another pass to remove dead code and improve idioms.

Finally, I'm glad to see the `'mw` lifetime go away but it's not obvious to me why `<D: '_, '_>` is used.  Is this inherited from `hyper` changes?

**Comment by GitCop** [comment]
Thanks for contributing! Unfortunately, I'm here to tell you there were style issues with your Pull Request:

 * Commit: 148f1131eef783bacd1b94382071995b2e5219a3
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 7f01f95114b56534316bae051c786bf1dfa4cdb0
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 4cd09ccd0e3e3c77413c1288bc48e127716fc0d4
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 264cd23e512606dc07d736702d588dc1c34a3883
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 362e4779517836c1b36011a9285d81b2999b0364
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 8a828acb8ddf17187059c1d8f9fd7fa6f0029e18
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 8293706cbc111da247e85143594b13885401a6b0
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: ea3d8f13c6bf6b6635178904ddaab52b14951c7f
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 8ff8981f43f73017083b4ad014909109085bf2d9
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: df8b94f35a2f70efae075eced8e6eef817a0e3e8
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: ad183a6ffbf97daeeac3b950b8571461a6c85956
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 6896d75e5193bfd7aac2655cfaf65e81ec16ffc8
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 63a7723cab97c4015239c1d83f5081547f6758da
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 4cb386ac04512e113d505b4bebd37ea0ec2780b7
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 84038c45d7a384f0a5141c78e54f2e15b6246ea6
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: ee96b9f4f9f0f3b030c61b81bf721c4688c39292
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 399da651c6c58261f8cc477fa132f9ae32261378
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 52ec773ed4f4ace52148602182c6ad26393eabd6
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 51c1f51aec5e3da0ab39028d4599d124e8b57c66
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 72409b21ef3c606d49a73b7807e0c5402645481d
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: e13cfe6bb3f073bb60362d44c51f86c1d8d3b56a
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 4bd354474499dd1df0aada3fc6af83892566c4b9
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: fe20927654a7958f56ce7ed5a1536d4542463900
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: e90ed6f5992ebad01ec46ae4032d0265d7cdc9e8
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 5a7524dd3d573da456da9eb37fef7a38b902c84e
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 75daf1bd34f9bde78957181f19fd3369a88dfd27
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**
  * Invalid type. Valid types are **feat, fix, docs, style, refactor, perf, test, chore, revert, doc**

 * Commit: 6495879c7378af46fcfc2f7a0634e9a0f2e71dff
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: ee8b8132d03980f4d1e3e237d72aa40ef934eeae
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: c00e64c7e1accccfdfe256d7bb302d4637b5d811
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 3c91d388e98dd6490851b94babfbd21a2df1e3b4
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/nickel-org/nickel.rs/blob/master/contributing.md

---
This message was auto-generated by https://gitcop.com

File: examples/integration_testing.rs [hunk 1]
```diff
@@ -87,35 +114,37 @@ where D: Database {
...
-            None => {
-                serde_json::from_str::<serde_json::Value>(&format!(
-                    r#"{{ "message": "Hello {}, I don't know your age" }}"#,
-                    client.name
-                )).unwrap()
-            }
-        }
-    });
+    server.post("/", JsonPost);
+    // TODO: the middleware macro has not yet been updated to support async
```
[the remark this answers is not in the document]
Comment (jolhoeft) [reply examples/integration_testing.rs:118]: I want to defer the final decision on the middleware macro. With the new version of hyper we may be able to eliminate it completely in favor of simple closures.

File: examples/json.rs [hunk 2]
```diff
@@ -1,33 +1,50 @@
...
-        middleware! { |request, response|
-            let person = try_with!(response, {
-                request.json_as::<Person>().map_err(|e| (StatusCode::BadRequest, e))
-            });
-            format!("Hello {} {}", person.first_name, person.last_name)
-        },
-    );
+    server.post("/", JsonPost);
+
+    // TODO: the middleware macro has not yet been updated to support async
```
[the remark this answers is not in the document]
Comment (jolhoeft) [reply examples/json.rs:38]: Ditto, defering

File: examples/no_macro_response_custom_data.rs [hunk 3]
```diff
@@ -1,9 +1,7 @@
-extern crate nickel;
-extern crate plugin;
-extern crate typemap;
+// TODO: extensions support has been (hopefully) temporarily removed due to it
+// not working in the async environment
 
 use nickel::{Nickel, HttpRouter, Request, Response, MiddlewareResult};
-use plugin::Extensible;
```
[the remark this answers is not in the document]
Comment (jolhoeft) [reply examples/no_macro_response_custom_data.rs:6]: May not have been used

File: src/body_parser.rs [hunk 4]
```diff
@@ -1,15 +1,18 @@
+// WARNING: this module is no longer used, and is only being kept around for
+// documentaion as part of migration to async.
```
[the remark this answers is not in the document]
Comment (jolhoeft) [reply src/body_parser.rs:2]: Typo fixed. The module itself isn't currently built (and fails). Once the async migration is complete, we'll just delete it.

File: src/response.rs [hunk 5]
```diff
@@ -199,125 +207,141 @@ impl<'a, D> Response<'a, D, Fresh> {
...
+    pub fn bail<T>(self, message: T) -> MiddlewareResult<D>
             where T: Into<Cow<'static, str>> {
         let _ = self.end();
         unsafe { Err(NickelError::without_response(message)) }
     }
 
     /// Flushes all writing of a response to the client.
+    // TODO: migration cleanup
+    //
+    // Should be easy, may not even be needed
```
[the remark this answers is not in the document]
Comment (jolhoeft) [reply src/response.rs:299]: Will clean up in the futre

File: src/response.rs [hunk 6]
```diff
@@ -335,78 +359,20 @@ fn matches_content_type () {
...
-        LastModified,
-        Location,
-        Pragma,
-        Referer,
-        Server,
-        SetCookie,
-        TransferEncoding,
-        Upgrade,
-        UserAgent,
-        Vary
```
[the remark this answers is not in the document]
Comment (jolhoeft) [reply src/response.rs:410]: Hyper made a major change to how headers are handled, so this is no longer relevant.

File: src/server.rs [hunk 7]
```diff
@@ -1,114 +1,83 @@
...
-        server.keep_alive(keep_alive_timeout);
-
-        let listening = match thread_count {
-            Some(threads) => server.handle_threads(arc, threads),
-            None => server.handle(arc),
-        };
+    pub async fn serve<A: ToSocketAddrs>(self,
+                                         addr: A,
+                                         keep_alive_timeout: Option<Duration>, // TODO: migration cleanup - use this
+                                         thread_count: Option<usize>) // TODO: migration cleanup - use or remove this
```
[the remark this answers is not in the document]
Comment (jolhoeft) [reply src/server.rs:35]: Not at this stage, this is a bare bones migration. Just enough to get basic functionality working. First PR of several.

File: src/server.rs [hunk 8]
```diff
@@ -1,114 +1,83 @@
...
+            let data = self.shared_data.clone();
+            let res_templates = self.templates.clone();
+            async move {
+                Ok::<_, Infallible>(service_fn(move |req: Request<Body>| {
+                    let mw2 = mw.clone();
+                    let req_data2 = data.clone();
+                    let res_data2 = data.clone();
+                    let res_templates2 = res_templates.clone();
+                    async move {
+                        let res = Response::builder().status(StatusCode::NOT_FOUND).body(Body::empty()).unwrap();
```
[the remark this answers is not in the document]
Comment (jolhoeft) [reply src/server.rs:51]: This is the default response. If no middle updates the response, a 404 is returned.

File: src/server.rs [hunk 9]
```diff
@@ -1,114 +1,83 @@
...
+                        let final_res = mw2.invoke(nickel_req, nickel_res).await;
+                        Ok::<_, Infallible>(final_res)
+                    }
+                }))
+            }
+        });
+        let server = HyperServer::bind(&socket_addr).serve(make_svc);
 
-        listening.map(ListeningServer)
+        println!("Listening on http://{}", socket_addr);
```
[the remark this answers is not in the document]
Comment (jolhoeft) [reply src/server.rs:66]: Probably, but deferring for the future

**Comment by cburgdorf** [comment]
Thank you so much for your continued support 🙏♥️

**Comment by jolhoeft** [comment]
You're welcome. A lot to be done before a 0.12 can release, though.

**Pull request merged**